src/index.tsx

```diff
@@ -0,0 +1,172 @@
+    style = {},
+    canvasStyle = {},
+    onStart = undefined,
+    onAnimate = () => {
+      console.log('animating')
+    },
+    onResize = undefined,
+    onAction = undefined,
+    tempo = undefined,
+    spaceRef: propsSpaceRef = undefined,
+    formRef: propsFormRef = undefined
+  }: PtsCanvasProps,
+  ref: ForwardedRef<HTMLCanvasElement>
+) => {
+  // Set canvRef to be either the forwarded ref if its a MutableRefObject, or our own local ref otherwise
+  const canvRef = ref && typeof ref !== 'function' ? ref : useRef(null)
+  const spaceRef = propsSpaceRef ?? useRef<CanvasSpace>()
+  const formRef = propsFormRef ?? useRef<CanvasForm>()
+
+  /**
+   * When canvRef Updates (ready for space)
+   */
+  useEffect(() => {
+    if (!canvRef || !canvRef.current) return
+    // Create CanvasSpace with the canvRef and assign to spaceRef
+    // Add animation, tempo, and play when ready (call back on CanvasSpace constructor)
+    spaceRef.current = new CanvasSpace(canvRef.current).setup({
+      bgcolor: background,
+      resize,
+      retina
+    })
+
+    // Assign formRef
+    formRef.current = spaceRef.current.getForm()
+
+    // By having individual handler props, we can expose what we need to the
+    // underlying functions, like our Form instance
+    spaceRef.current.add({
+      start: (bound: Bound) => {
+        onStart && onStart(bound, spaceRef.current, formRef.current)
+      },
+      animate: (time?: number, ftime?: number) => {
+        onAnimate && onAnimate(spaceRef.current, formRef.current, time, ftime)
+      },
+      resize: (bound: Bound, event: Event) => {
+        // eslint-disable-line no-undef
+        onResize && onResize(spaceRef.current, formRef.current, bound, event)
+      },
+      action: (type: string, px: number, py: number, evt: Event) => {
+        // eslint-disable-line no-undef
+        onAction &&
+          onAction(spaceRef.current, formRef.current, type, px, py, evt)
+      }
+    })
+
+    // Add tempo if provided
+    if (tempo) {
+      spaceRef.current.add(tempo)
+    }
+
+    // Return the cleanup function (similar to ComponentWillUnmount)
+    return () => {
+      spaceRef.current && spaceRef.current.dispose()
+    }
+  }, [canvRef])
+
+  /**
+   * When Touch updates
+   */
+  useEffect(() => {
+    spaceRef.current && spaceRef.current.bindMouse(touch).bindTouch(touch)
+  }, [touch])
+
+  /**
+   * Play or stop based on play prop
+   * */
+  const maybePlay = () => {
+    if (play) {
+      spaceRef.current && spaceRef.current.play()
+    } else {
+      spaceRef.current && spaceRef.current.playOnce(0)
+    }
+  }
+
+  /**
+   * When anything updates
+   */
+  useEffect(() => {
+    maybePlay()
+  })
+
+  return (
+    <div className={name || ''} style={style}>
+      <canvas
+        className={name ? name + '-canvas' : ''}
+        ref={canvRef}
+        style={canvasStyle}
+      />
+    </div>
+  )
+}
+
+export const PtsCanvas = forwardRef<HTMLCanvasElement, PtsCanvasProps>(
```

Comment:
aren't HTMLCanvasElement and PtsCanvasProps in the opposite order?

Reply:
Haha I thought the same thing, but no they are reversed for some reason, https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/forward_and_create_ref/

Ref type first, then props type.